sw/otbn/code-snippets/modexp.S

```diff
@@ -1288,6 +1288,142 @@ modexp:
+ *                              squared Montgomery Modulus RR mod M
+ *   @param[in]  dmem[44] dptr_a_sqr: dmem pointer to first limb of result C
+ *   @param[in]  dmem[48] dptr_b_sqr: dmem pointer to first limb of result C
+ *   @param[in]  dmem[52] dptr_c_sqr: dmem pointer to first limb of result C
+ *   @param[in]  dmem[56] N: Number of limbs per bignum
+ *   @param[in]  dmem[60] N-1: Number of limbs per bignum minus 1
+ *
+ * third descriptor used for multiplication:
+ *   @param[in]  dmem[64] dptr_M: dmem pointer to first limb of modulus M
+ *   @param[in]  dmem[68] dptr_m0d: dmem pointer to Montgomery Constant m0'
+ *   @param[in]  dmem[72] dptr_RR: dmem pointer to first limb of
+ *                              squared Montgomery Modulus RR mod M
+ *   @param[in]  dmem[76] dptr_a_mul: dmem pointer to first limb of base A
+ *   @param[in]  dmem[80] dptr_b_mul: dmem pointer to first limb of result C
+ *   @param[in]  dmem[84] dptr_c_mul: dmem pointer to first limb of result C
+ *   @param[in]  dmem[88] N: Number of limbs per bignum
+ *   @param[in]  dmem[92] N-1: Number of limbs per bignum minus 1
+ *
+ * fourth descriptor used for back-conversion:
+ *   @param[in]  dmem[96] dptr_M: dmem pointer to first limb of modulus M
+ *   @param[in]  dmem[100] dptr_m0d: dmem pointer to Montgomery Constant m0'
+ *   @param[in]  dmem[104] dptr_RR: dmem pointer to first limb of
+ *                              squared Montgomery Modulus RR mod M
+ *   @param[in]  dmem[108] dptr_a_ex1: dmem pointer to first limb of base A
+ *   @param[in]  dmem[116] dptr_c_ex1: dmem pointer to first limb of result C
+ *   @param[in]  dmem[120] N: Number of limbs per bignum
+ *   @param[in]  dmem[124] N-1: Number of limbs per bignum minus 1
+ *
+ * clobbered registers: x3 to x13, x16 to x31
+ *                      w0 to w3, w24 to w30
+ *                      w4 to w[4+N-1]
+ * clobbered Flag Groups: FG0, FG1
+ */
+modexp_65537:
+  /* convert to montgomery domain montmul(A,RR)
+  in = montmul(A,RR) = C*R mod M */
+  jal       x1, mulx
+
+  /* pointer to out buffer */
+  lw        x21, 116(x0)
+
+  /* zeroize w2 and reset flags */
+  bn.sub    w2, w2, w2
+
+  /* this loop initializes the output buffer with -M */
+  loop      x30, 3
+    /* load limb from modulus */
+    bn.lid    x11, 0(x16++)
+
+    /* subtract limb from 0 */
+    bn.subb   w2, w31, w2
+
+    /* store limb in dmem */
+    bn.sid    x11, 0(x21++)
+
+  /* 65537 = 0b10000000000000001
+               ^ sqr + mult
+    out = montmul(out,out)       */
+  jal       x1, sqrx_exp
+
+  /* out = montmul(in,out)       */
+  jal       x1, mulx_exp
+
+  /* store multiplication result in output buffer */
+  li        x8, 4
+  loop      x30, 2
+    /* store selected limb to dmem */
+    bn.sid    x8, 0(x21++)
+    addi      x8, x8, 1
+
+  /* 65537 = 0b10000000000000001
+                ^<< 16 x sqr >>^   */
+  loopi      16, 2
+    /* square: out = montmul(out, out) */
+    jal       x1, sqrx_exp
+    nop
```

Comment:
This "jump inside a loop" precludes loop implementations based on counting instructions, which I think we were considering at one point. @GregAC: is this ok with your planned implementation?

Also, I wonder whether we can avoid needing the `nop` (does the loop match check look at link addresses?). If not, we'll have to document that quite carefully.

Reply:
I am afraid the jump-in-loop feature is sth. we really need. We are using this a lot for multiplications and the Montgomery loop.

Regarding `nop`: I thought our conclusion from #3609 was a bit that we need it. But I agree we should definitely document this and probably have the tool chain throw a warning if last instruction in a loop body is a `JAL`

Reply:
My current RTL will work with branches/jumps in loops provided they're not the final instruction of a loop body.

Branches/jumps in a loop aren't an inherent problem, just opens the door to weird behaviour the hardware will have a hard time detecting. As it's all hand-crafted assembly we're probably fine with odd behaviour happening if you do something weird/stupid (e.g. branch from loop body then never branch back or branch back above the and start it again), indeed we could probably come up with some kind of static analysis tool to detect the dodgy cases and/or add some checking in the simulator for it.

The real question is around whether this is a security concern and whether we'd want hardware to detect odd behaviour and halt with an error.

I've got a big list of loop corner cases to go over, so let's maybe hold this discussion until I've got that out.